### PR TITLE
Onboard validator's Beacon REST API usage to e2e tests

### PR DESCRIPTION
**Pull request by PatriceVignola**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

It onboards the validator's Beacon REST API usage to the e2e tests by going through the following steps:
1. Convert the convert the `use_beacon_api` buildtime flag to a `--enable-beacon-rest-api` flag
2. Modify the e2e test evaluators to take a `validatorHelpers.NodeConnection` as an argument instead of a `*grpc.ClientConn`. There's unfortunately not a clean way to do this without touching a bunch of files, but the change is actually simple and `validatorHelpers.NodeConnection` is simply a wrapper around `*grpc.ClientConn`. The behavior of existing e2e tests will not change.
3. Add tests to the presubmit collateral that turn on the `--enable-beacon-rest-api` flag.

### CODE REVIEW

File: testing/endtoend/endtoend_test.go [hunk 1]
```diff
@@ -132,11 +133,11 @@ func (r *testRunner) scenarioRunner() {
 	r.runBase([]runEvent{r.scenarioRun})
 }
 
-func (r *testRunner) waitExtra(ctx context.Context, e types.Epoch, conn *grpc.ClientConn, extra types.Epoch) error {
+func (r *testRunner) waitExtra(ctx context.Context, e types.Epoch, conn validatorHelpers.NodeConnection, extra types.Epoch) error {
```
Comment (nisdas) [inline testing/endtoend/endtoend_test.go:136]: Please revert this

File: testing/endtoend/endtoend_test.go [hunk 2]
```diff
@@ -174,7 +175,7 @@ func (r *testRunner) waitForChainStart() {
 }
 
 // runEvaluators executes assigned evaluators.
-func (r *testRunner) runEvaluators(conns []*grpc.ClientConn, tickingStartTime time.Time) error {
+func (r *testRunner) runEvaluators(conns []validatorHelpers.NodeConnection, tickingStartTime time.Time) error {
```
Comment (nisdas) [inline testing/endtoend/endtoend_test.go:178]: revert this change

**Review by nisdas — CHANGES_REQUESTED** [review]
There is a misunderstanding on how grpc is used in our evaluators. All these evaluators query the `beacon-node` not the `validator` . So there is no need to change the format of the connection to `validatorHelpers.NodeConnection` . The evaluators will still work as expected because they query the beacon node, and whether the validator is running via grpc or the beacon rest-api makes no difference to this.

File: testing/endtoend/endtoend_test.go (hunk 2 above)
Comment (nisdas) [reply testing/endtoend/endtoend_test.go:178]: @PatriceVignola That is just using the grpc client to access the related rpc methods from the beacon node. They do not access the validator Ex:
```
StreamBlocks , StreamBlocksAltair
```

are methods that are served by the beacon node. Whether you are running with the rest-api or not makes no difference here.

Comment (PatriceVignola) [reply testing/endtoend/endtoend_test.go]: Ah I see what you mean now. I was confused by the calls to `GetDuties()` and `DomainData()` in the `slashing.go` file, which both are endpoints that we intend to convert to REST. Sorry for the confusion! I went ahead and reverted the changes to the evaluators, it should look much better now.

File: testing/endtoend/components/validator.go [hunk 3]
```diff
@@ -214,13 +220,15 @@ func (v *ValidatorNode) Start(ctx context.Context) error {
 		fmt.Sprintf("--%s=%d", flags.MonitoringPortFlag.Name, e2e.TestParams.Ports.ValidatorMetricsPort+index),
 		fmt.Sprintf("--%s=%d", flags.GRPCGatewayPort.Name, e2e.TestParams.Ports.ValidatorGatewayPort+index),
 		fmt.Sprintf("--%s=localhost:%d", flags.BeaconRPCProviderFlag.Name, beaconRPCPort),
+		fmt.Sprintf("--%s=http://localhost:%d", flags.BeaconRESTApiProviderFlag.Name, beaconRestApiPort),
```
Comment (nisdas) [inline testing/endtoend/components/validator.go:223]: I don't think we should be applying the flag here if the beacon rest api has not been enabled for the particular test case, ideally we should gate this to only be used if the `WithValidatorRESTApi` has been used.

File: testing/endtoend/components/validator.go [hunk 4]
```diff
@@ -187,6 +187,12 @@ func (v *ValidatorNode) Start(ctx context.Context) error {
 		beaconRPCPort = e2e.TestParams.Ports.PrysmBeaconNodeRPCPort
 	}
 
+	beaconRestApiPort := e2e.TestParams.Ports.PrysmBeaconNodeGatewayPort + index
+	if beaconRestApiPort >= e2e.TestParams.Ports.PrysmBeaconNodeGatewayPort+e2e.TestParams.BeaconNodeCount {
```
Comment (nisdas) [inline testing/endtoend/components/validator.go:191]: Remove this check here, we should always have validators pointing to the correctly allocated port. If the beacon api port leads to some non-existent being used, its better to fail early rather than invisibly.

File: testing/endtoend/minimal_scenario_e2e_test.go [hunk 5]
```diff
@@ -26,3 +26,24 @@ func TestEndToEnd_ScenarioRun_EEOffline(t *testing.T) {
 	runner.config.EvalInterceptor = runner.eeOffline
 	runner.scenarioRunner()
 }
+
+func TestEndToEnd_MultiScenarioRun_ValidatorRESTApi(t *testing.T) {
```
Comment (nisdas) [inline testing/endtoend/minimal_scenario_e2e_test.go:30]: I think running so many test cases for just the `REST` Api is going to lengthen our test times considerably. Currently the scenario tests take 3 hours to run, with all these additional tests they maybe be pushed up to 6 hours, which isn't really acceptable imo. I think having `TestEndToEnd_MinimalConfig_ValidatorRESTApi` should be enough, @rkapka what do you think ?

File: testing/endtoend/minimal_e2e_test.go [hunk 6]
```diff
@@ -9,3 +9,7 @@ import (
 func TestEndToEnd_MinimalConfig(t *testing.T) {
 	e2eMinimal(t, types.WithCheckpointSync()).run()
 }
+
+func TestEndToEnd_MinimalConfig_ValidatorRESTApi(t *testing.T) {
```
Comment (nisdas) [inline testing/endtoend/minimal_e2e_test.go:13]: Please shift this to the `minimal_scenario_e2e_test.go` file

File: testing/endtoend/minimal_scenario_e2e_test.go (hunk 5 above)
Comment (rkapka) [reply testing/endtoend/minimal_scenario_e2e_test.go:34]: I agree

Comment (PatriceVignola) [reply testing/endtoend/minimal_scenario_e2e_test.go]: I removed all tests except `TestEndToEnd_MinimalConfig_ValidatorRESTApi`

**Review by nisdas — APPROVED** [review]

**Review by rkapka — APPROVED** [review]
